### PR TITLE
Expose source ranges from errors for use with LSP

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Currently compiler errors are converted directly from the underlying error types into a Pretty error message.
The LSP server needs the converted pretty message, but it also needs to know where to display each message.
This change makes the error printer also return the relevant ranges where the error message applies.

This change will be needed by the LSP, but I'm making a separate PR with the intention of getting this chunk merged ASAP to avoid merge conflicts since the lsp branch might be longer lived.

I recommend reviewing with white space diffs hidden

## Implementation notes

* Splits `prettyParseError` up into a lower-level `renderParseErrors` which renders each individual error message alongside the ranges each message applies at (each parse error may result in multiple error messages and ranges). We need this to get a handle on the individual messages and their ranges before they get squashed together for printing to screen so that we can render each message at the appropriate spot in the editor.
* Change the error return value of `parseAndSynthesizeFile` from just the Names of the failed typecheck to the full UnisonFile. The LSP can still extract useful information from the UnisonFile even if it failed to typecheck, and source annotations are lost when converting to a Names. Everywhere that needed the Names can simply extract the names from the UnisonFile.
* Split off the codepath for typechecking a scratchfile into its own command which we can call from the LSP

## Test coverage

Transcripts are all passing and unchanged (except a tiny formatting fix I made to one of the errors) 👍🏼

- - -
This change is [Reviewable](https://reviewable.io/reviews/unisonweb/unison/3267)

### CODE REVIEW

File: parser-typechecker/src/Unison/PrintError.hs [hunk 1]
```diff
@@ -1115,10 +1115,26 @@ prettyParseError ::
...
+prettyParseError s e =
+  mconcat (fst <$> renderParseErrors s e) <> lexerOutput
+  where
+    lexerOutput :: Pretty (AnnotatedText a)
+    lexerOutput =
+      if showLexerOutput
+        then "\nLexer output:\n" <> fromString (L.debugLex' s)
+        else mempty
+
+renderParseErrors ::
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/PrintError.hs:1127]: New combinator that renders a parser error into a list of errors and the ranges within the file to which each of those errors applies.

These ranges will be used to tell the LSP where to show diagnostic errors.

File: parser-typechecker/src/Unison/PrintError.hs [hunk 2]
```diff
@@ -1115,10 +1115,26 @@ prettyParseError ::
...
+  forall v.
+  Var v =>
+  String ->
+  Parser.Err v ->
+  [(Pretty ColorText, [Range])]
+renderParseErrors s = \case
+  P.TrivialError _ (LexerError ts e) _ -> [(go e, ranges)]
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/PrintError.hs:1134]: The diffs in this function look noisy, but if you turn off whitespace diffs you'll see it's mostly just re-arranging things into a let binding and adding the appropriate ranges for each error.

File: unison-cli/src/Unison/Codebase/Editor/HandleCommand.hs [hunk 3]
```diff
@@ -282,6 +282,20 @@ commandLine config awaitInput setBranchRef rt sdbxRt notifyUser notifyNumbered l
...
+typecheckCommand ::
+  (Random.DRG gen, Monad m) =>
+  Codebase m Symbol Ann ->
+  [Type Symbol Ann] ->
+  NamesWithHistory.NamesWithHistory ->
+  SourceName ->
+  LexedSource ->
+  gen ->
+  m (TypecheckingResult Symbol)
+typecheckCommand codebase ambient names sourceName source gen = do
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleCommand.hs]: Pulling this out into a separate command allows us to call the same typechecker path in both UCM and the LSP.

**Review by aryairani — APPROVED** [review]
> Reviewed 6 of 11 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/unisonweb/unison/3267)* status: all files reviewed, 3 unresolved discussions (waiting on @mitchellwrosen and @tstat)

File: parser-typechecker/src/Unison/PrintError.hs [hunk 4]
```diff
@@ -1400,145 +1434,173 @@ prettyParseError s = \case
...
+                if missing
+                  then "Make sure it's spelled correctly."
+                  else "Try hash-qualifying the type you meant to reference."
+              ]
+       in (msg, [rangeForToken tok])
       where
         missing = Set.null referents
     go (Parser.ResolutionFailures failures) =
-      Pr.border 2 . prettyResolutionFailures s $ failures
+      -- TODO: We should likely output separate diagnostics for each failure.
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/PrintError.hs]: Expand this comment

**Pull request merged**